internal/service/resources/collector/collector.go

```diff
@@ -0,0 +1,229 @@
+		slog.Debug("collecting data from data source", "source", d)
+		if err := c.executeOneDataSource(ctx, d); err != nil {
+			slog.Warn("failed to collect data from data source", "source", d, "error", err)
+		} else {
+			slog.Debug("collected data from data source", "source", d)
+		}
+	}
+	slog.Debug("collector loop complete", "sources", len(c.dataSources))
+}
+
+// executeOneDataSource runs a single iteration of the main loop for a specific data source instance.
+func (c *Collector) executeOneDataSource(ctx context.Context, dataSource DataSource) (err error) {
+	// Increment the generation id
+	dataSource.IncrGenerationID()
+
+	// Get the list of resource pools for this data source
+	pools, err := c.collectResourcePools(ctx, dataSource)
+	if err != nil {
+		return fmt.Errorf("failed to collect resource pools: %w", err)
+	}
+
+	// Get the list of resources for this data source
+	_, err = c.collectResources(ctx, dataSource, pools)
+	if err != nil {
+		return fmt.Errorf("failed to collect resources: %w", err)
+	}
+
+	// TODO: purge stale record
+
+	// TODO: persist data source info
+
+	return nil
+}
+
+// collectResources collects Resource objects from the data source, persists them to the database,
+// and signals any change events to the notification processor.
+func (c *Collector) collectResources(ctx context.Context, dataSource DataSource,
+	pools []models.ResourcePool) ([]models.Resource, error) {
+	resources, err := dataSource.GetResources(ctx, pools)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resources: %w", err)
+	}
+
+	// Loop over the set of resources and create the associated resource types.
+	seen := make(map[uuid.UUID]bool)
+	for _, resource := range resources {
+		resourceType, err := dataSource.MakeResourceType(&resource)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make resource type from '%v': %w", resource, err)
+		}
+
+		if _, ok := seen[resourceType.ResourceTypeID]; ok {
+			// We have already seen this one so skip
+			continue
+		}
+		seen[resourceType.ResourceTypeID] = true
+
+		dataChangeEvent, err := persistObjectWithChangeEvent(
+			ctx, c.repository.Db, *resourceType, resourceType.ResourceTypeID, nil, func(object interface{}) any {
+				record, _ := object.(models.ResourceType)
+				return models.ResourceTypeToModel(&record)
+			})
+		if err != nil {
+			return nil, fmt.Errorf("failed to persist resource type': %w", err)
+		}
+
+		if dataChangeEvent != nil { // nolint: staticcheck
+			// TODO: notify the notification processor of the new data change event via a channel
+		}
+	}
+
+	// Loop over the set of resources and insert (or update) as needed
+	for _, resource := range resources {
+		dataChangeEvent, err := persistObjectWithChangeEvent(
+			ctx, c.repository.Db, resource, resource.ResourceID, &resource.ResourcePoolID, func(object interface{}) any {
+				record, _ := object.(models.Resource)
+				return models.ResourceToModel(&record, nil)
+			})
+		if err != nil {
```

Comment:
Same comment as in the resource pool for errors.

Let's try to keep the DB as consistent as possible.

Say one of these fails. We will have commited resource pools, resource types and some or at the worst none resources (if the first fails). The whole data willl be inconsistent until the next sync.

Am I right?

Reply:
I'll give this some thought based on how things behave in testing at larger scale.  Like I said in the previous comment there shouldn't be any reason for one record to pass and another to fail so that leaves db connection issues which should be quite rare.  There is too much data to try and do everything under one single transaction so it would be difficult to ensure the data is always consistent.